basic/digital-library/src/lib.rs

```diff
@@ -0,0 +1,194 @@
+use sbor::*;
+use scrypto::prelude::*;
+
+#[derive(TypeId, Encode, Decode)]
+struct Book {
+    title: String,
+    author: String,
+}
+
+#[derive(TypeId, Encode, Decode)]
+struct BorrowedBook {
+    // when the user borrowed the book
+    epoch: u64,
+    user_id: Address,
+}
+
+blueprint! {
+    struct Library {
+        // librarian (admin) badge
+        librarian_badge_def: ResourceDef,
+
+        // holds the library funds, from late fees and memberships
+        fees: Vault,
+
+        // maps ISBN numbers to books
+        books: HashMap<String, Book>,
+
+        // member badge
+        member_badges: Vault,
+        member_badge_def: ResourceDef,
+        // membership cost in XRD
+        membership_price: Decimal,
+
+        // maps ISBN numbers to membership badge of borrower
+        borrowed_books: HashMap<String, BorrowedBook>,
+        // the number of epochs books can be borrowed for
+        borrow_epochs: u64
+    }
+
+    impl Library {
+        // the library requires:
+        // - member_badge_count: number of members allowed to join at once
+        // - membership_price: the cost in XRD to join the library
+        // - borrow_epochs: the number of epochs books are borrowed for
+        // the librarian badge is returned to the caller
+        pub fn new(member_badge_count: u32, membership_price: Decimal, borrow_epochs: u64) -> (Component, Bucket) {
+            let librarian_badge_bucket = ResourceBuilder::new()
+                .metadata("name", "Librarian Badge")
+                .metadata("symbol", "LB")
+                .new_badge_fixed(1);
+
+            let mut books = HashMap::new();
+            let book_one = Book { author: String::from("James S. A. Corey"), title: String::from("Leviathan Wakes") };
+            let book_two = Book { author: String::from("Frank Herbert"), title: String::from("Dune") };
+            let book_three = Book { author: String::from("Dan Abnett"), title: String::from("Horus Rising") };
+            books.insert(String::from("9781611297560"), book_one);
+            books.insert(String::from("9780450011849"), book_two);
+            books.insert(String::from("9781844162949"), book_three);
+
+            let member_badges_bucket = ResourceBuilder::new()
+                .metadata("name", "Library Membership Badge")
+                .metadata("symbol", "LMB")
+                .new_badge_fixed(member_badge_count);
```

Comment:
I think it would be cool for this badge to be mutable. You would create another single badge "member_badge_minter", stored inside a vault of the component. Then inside `register`, you mint a new badge only when a user needs to register. I think this will make it more optimized to get the total number of members. Instead of having to count each keys in the hashmap, you could do `member_badges_resource_def.supply()`